feat: extend modification move instruction to support `sourceElementInstanceKey`

mustafadagher opened the pull request:
## Summary
Extend process instance modification “move” to support identifying the source by its element instance key via a new field, **sourceElementInstanceKey**, in addition to the existing element ID–based approach. This enables precise moves when multiple instances of the same element are active (e.g., multi-instance activities or parallel branches), while also improving how the engine determines the correct ancestor scope for the target..

### What’s changed

#### Engine

* **Move-by-instance-key support**

  * Move instructions can now specify the source by **element instance key** via `sourceElementInstanceKey` in the `MoveInstruction` record.
  * The processor validates move-by-key instructions first (existence and process ownership) and then maps them to terminate/activate operations, similar to ID-based moves.
  * The processor **rejects** move instructions that specify **both** a `sourceElementInstanceKey > 0` and a non-blank `sourceElementId`, for consistency with terminate instructions and to avoid surprising behavior.

* **Ancestor scope resolution**

  * Introduced logic to **infer the ancestor scope from the source hierarchy** when `inferAncestorScopeFromSourceHierarchy` is enabled on the move instruction.
  * The engine walks up the source element instance’s ancestry and finds the first instance whose element is in the target element’s flow scope chain. The new element instance is then activated in this inferred scope, ensuring the target is created in the correct flow scope even for deeply nested or complex models.
  * Existing behavior for **direct ancestor** moves (explicit `ancestorElementInstanceKey`) remains supported.

* **Validation and error handling**

  * Move instructions are rejected if:
    * The source element instance key does not exist.
    * The source element instance belongs to a different process instance.
    * Both `sourceElementInstanceKey` and `sourceElementId` are set.
  * Error messages were updated to clearly describe invalid combinations and ownership violations.

#### Protocol / Records / Gateway

* Extended `MoveInstruction` with a `sourceElementInstanceKey` field (default `-1` when not used).
* Renamed the flag that controls inferred ancestor resolution from `useSourceParentKeyAsAncestorScopeKey` to **`inferAncestorScopeFromSourceHierarchy`**, reflecting the actual behavior (inferring the ancestor from the source’s hierarchy rather than simply using the immediate parent).
* Exposed `getSourceElementInstanceKey()` on the process instance modification record value interface.

#### REST

* Modelled the move source as a dedicated **`SourceElementInstruction`** that can represent:
  * **byId**: `sourceElementId`
  * **byKey**: `sourceElementInstanceKey`
* Modelled the ancestor scope as an **`AncestorScopeInstruction`** that can represent:
  * **direct**: explicit `ancestorElementInstanceKey`
  * **inferred**: infer ancestor scope from source hierarchy
* Updated the OpenAPI definitions and client-side models accordingly to avoid ambiguous oneOfs and to prevent null/unused fields from being serialized.

#### Java client API

Extended the Java client’s **modify process instance** command with new overloads and updated naming to better reflect the engine semantics:

* **Move by element ID**

  * `moveElements(String sourceElementId, String targetElementId, long ancestorElementInstanceKey)`  
    Creates a move instruction where the target is activated in the **explicit ancestor** scope.

  * `moveElementsWithInferredAncestor(String sourceElementId, String targetElementId)`  
    Creates a move instruction where the target’s ancestor scope is **inferred from the source element’s hierarchy**, using the same algorithm as the engine.

* **Move by element instance key**

  * `moveElement(long sourceElementInstanceKey, String targetElementId)`  
    Creates a move instruction that terminates the given source instance and activates a new one at the target element. The target is created in an existing flow scope instance when possible.

  * `moveElement(long sourceElementInstanceKey, String targetElementId, long ancestorElementInstanceKey)`  
    Same as above, but with an **explicit ancestor** instance key controlling where the target is created.

  * `moveElementWithInferredAncestor(long sourceElementInstanceKey, String targetElementId)`  
    Same as above, but the ancestor scope is **inferred from the source element’s hierarchy**, matching the engine’s inferred-ancestor behavior.

Internally, these methods now construct the appropriate REST **`ProcessInstanceModificationMoveInstruction`** using the new `SourceElementInstruction` (`sourceType = "byId"` or `"byKey"`) and `AncestorScopeInstruction` (`ancestorScopeType = "direct"` or `"inferred"`).

#### Exporters

* Updated the **Elasticsearch** and **OpenSearch** templates for process instance modification records to include the new `sourceElementInstanceKey` field with the correct JSON field name, ensuring it is indexed and searchable.

#### Tests

* Extended and refactored tests to cover:
  * Successful move-by-key scenarios across different scopes and multi-instance structures.
  * Rejections when:
    * The source element instance key is invalid or belongs to another process instance.
    * Both `sourceElementInstanceKey` and `sourceElementId` are set.
  * Correct ancestor scope inference for both ID-based and key-based moves.
* Updated golden records and JSON serialization tests to reflect the new protocol fields and structure.

BREAKING CHANGE:
In `gateway.proto`: the `MoveInstruction` message was updated to add `sourceElementInstanceKey` on top and rename `useSourceParentKeyAsAncestorScopeKey` to `inferAncestorScopeFromSourceHierarchy`, but this is acceptable since the original message is not released yet.

## Related issues

closes #41857

mustafadagher committed:
feat: extent protocol record to support moving a single element instance by key

mustafadagher committed:
feat: extent zeebe exporters templates to support moving a single element instance by key

mustafadagher committed:
feat: extend engine modification processor move instruction to support `sourceElementInstanceKey`

Previously, the modification move instruction only supported moving elements
by `sourceElementId`, which would find all element instances with that element
ID and terminate them, then activate the `targetElementId` for each.

This change extends the move instruction to also support `sourceElementInstanceKey`,
which allows moving a specific element instance by its key. This terminates only
that particular element instance and activates exactly one target element.

The implementation:
- Adds validation for `sourceElementInstanceKey` to check that the element
  instance exists and belongs to the correct process instance
- Handles move instructions by instance key separately from those by element ID
- Supports `ancestorScopeKey` and `useSourceParentKeyAsAncestorScope` options
- Supports variable instructions on move by instance key

mustafadagher committed:
test: covert terminate/activate pairs to move instructions

mustafadagher committed:
refactor: removed `Assertion.` prefix from `assertThat()`

mustafadagher committed:
fix: pass the correct intent to `BatchOperationCreationRecordValue` in client

Unrelated to this PR, but thought it's worth fixing

mustafadagher committed:
feat: find proper ancestor scope when moving between nested subprocesses

When using `isUseSourceParentKeyAsAncestorScope` to move an element from a
deeply nested subprocess to a less nested target element, the source's direct
parent key may not be a valid ancestor scope for the target element.

For example, moving from task1 (inside subprocess C, which is inside
subprocess B) to task2 (directly inside subprocess B) would incorrectly use
subprocess C's instance as the ancestor scope, but task2's flow scope is
subprocess B.

This change adds a helper method `findProperAncestorScopeKeyForTarget` that
traverses up the source element instance's ancestry to find an instance whose
element ID matches one of the target element's flow scope hierarchy.

mustafadagher commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationModifyProcessor.java:
```diff
@@ -265,12 +272,43 @@ public void processRecord(final TypedRecord<ProcessInstanceModificationRecord> c
+    final var moveInstructionsByInstanceKey =
+        moveInstructions.stream()
+            .filter(instruction -> instruction.getSourceElementInstanceKey() > 0)
+            .toList();
```
This was intentional, if a key is provided, the instruction is considered a move-by-key

mustafadagher replied:
I have a feeling that if you pass a `key` you know you want to terminate/move only one element. But I can adjust it here for consistency too.

mustafadagher committed:
feat: reject move instructions with both source element id and instance key

Add validation to reject process instance modification move instructions when both
source element id and source element instance key are provided. Only one of them
should be specified.

mustafadagher committed:
feat: extend PI modify move to support source element instance key to API protocol

mustafadagher committed:
feat: extend PI modify move to support source element instance key to gateway

mustafadagher committed:
feat: extend PI modify move to support source element instance key in Java SDK

mustafadagher committed:
test: validate move element with key E2E

tmetzke commented on zeebe/gateway-protocol/src/main/proto/v2/process-instances.yaml:
💭 The way we currently implement this in the engine, this here is not true anymore, if I'm not mistaken. I wonder if we should rename this to what we actually do in the engine. Correct me if I'm wrong, but I think what we do now is trying to use the target element's scope, right?

tmetzke reviewed (commented):
👍 Functionally, this is complete, as far as I can tell 🚀 
💭 There are potential improvements we can make regarding API correctness/clarity, code duplication, and target ancestor scope resolution options.

tmetzke commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationModifyProcessor.java:
```diff
@@ -265,12 +275,40 @@ public void processRecord(final TypedRecord<ProcessInstanceModificationRecord> c
+            .collect(
+                Collectors.toMap(
+                    ProcessInstanceModificationMoveInstructionValue::getSourceElementId,
+                    Function.identity()));
```
Yeah, I see what you mean 😅 In the end, it works. I would have to give it try myself to see if the combined one "looks" better or not 🤷 Maybe we can extract the stuff that we do for all move instructions into helper methods to reuse them. It's a taste thing, I would say, nothing mandatory in the end.

mustafadagher committed:
refactor: rename `UseSourceParentKeyAsAncestorScopeKey` to `InferAncestorScopeFromSourceHierarchy` across the stack

With `findProperAncestorScopeKeyForTarget` introduced in `ProcessInstanceModificationModifyProcessor`,
  we currently try to infer the closest common flow scope between the source and the target
  and use that as target ancestor scope

The rename reflects the change in behavior and removes confusion

mustafadagher committed:
refactor: use discriminated union for move instruction source element

Replace direct sourceElementId/sourceElementInstanceKey properties with a
SourceElementInstruction wrapper using discriminator pattern (byId/byKey).

- Remove custom deserializer in favor of OpenAPI discriminator support
- Simplify RequestMapper and validator logic
- Update Java SDK and tests for new API structure

mustafadagher committed:
fix: avoid serializing null values for required properties

Change JsonInclude from ALWAYS to USE_DEFAULTS for required properties
 in generated client code.
This prevents the client from sending null values
 for discriminated union properties (e.g., sourceElementId when
 sourceElementInstanceKey is set), which would fail gateway validation.

Compared to the original template file,
 We removed the ALWAYS include for "normal" fields
 to avoid sending `null` fields for requests with `oneOf` variants.
In the Camunda client, we currently generate classes
 that combine all fields of the alternatives.
 Thus, all required fields are merged into one request DTO.
As a result, all required fields are sent to the gateway,
 no matter if they are filled or not.
In the gateway, we're then receiving unexpected attributes,
 leading to Jackson errors upon deserialization.

tmetzke commented on clients/java/src/main/resources/templates/jackson_annotations.mustache:
```diff
@@ -0,0 +1,14 @@
+ * Else use custom behaviour, IOW use whatever is defined on the object mapper
+ }}
+  @JsonProperty(value = JSON_PROPERTY_{{nameInSnakeCase}}, required = {{#isNullable}}false{{/isNullable}}{{^isNullable}}{{required}}{{/isNullable}})
+  @JsonInclude({{#isMap}}{{#items.isNullable}}content = JsonInclude.Include.ALWAYS, {{/items.isNullable}}{{/isMap}}value = JsonInclude.Include.USE_DEFAULTS)
```
FYI, @nicpuppa and @megglos. If we see a better way to do this, we're happy to discuss.

nicpuppa replied:
I don't have another solution from the top of my head right now, so good to me. We can always think about another solution later

tmetzke commented on clients/java/src/test/java/io/camunda/client/tenant/SearchClientByTenantTest.java:
```diff
@@ -56,7 +56,7 @@ void shouldIncludeSortAndPaginationInSearchRequestBody() {
 
     assertThat(requestBody).contains("\"sort\":[{\"field\":\"clientId\",\"order\":\"DESC\"}]");
-    assertThat(requestBody).contains("\"page\":{\"limit\":2,\"after\":null,\"before\":null}");
+    assertThat(requestBody).contains("\"page\":{\"limit\":2}");
```
ℹ️ @nicpuppa and @megglos, this is a result of the change to the mustache template mentioned above.

tmetzke reviewed (approved):
👍 Looks good to me, really nice job, @mustafadagher 🚀 
🔧 There are some slight improvements we can make regarding Javadocs, but nothing blocking from my end. Pre-approving.
💭 If there are concerns regarding the approach with the overridden mustache template in the Camunda client, let's follow-up with @nicpuppa and @megglos.

tmetzke commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationModifyProcessor.java:
```diff
@@ -466,6 +501,118 @@ private void mapIdInstructions(
+   * @return the proper ancestor scope key, or -1 if no matching ancestor is found (will be handled
+   *     by the activation logic)
+   */
+  private long findProperAncestorScopeKeyForTarget(
```
@mustafadagher, can you follow-up with some frontend developers to assess if a sibling move instruction is easy to detect? In that case, we might want to (re)introduce an option to use the source's parent since that is less costly (we don't have to load the process model) than inferring the target scope.

mustafadagher replied:
@tmetzke I'll merge the PR as is for now, and will sync with @OmranAbazid about this, and re-introduce `useSourceAsParent` in a follow-up PR if needed.

tmetzke replied:
Awesome, thanks 👍
